### PR TITLE
EDSC-3750: Displaying additional browse imagery

### PR DESCRIPTION
**Pull request by trevorlang**
# Overview

### What is the feature?

Adds the ability to display multiple browse images in a carousel when focusing on a granule. Granules are visible in a thumbnail and can also be viewed in a larger modal.

### What is the Solution?

The `GranuleImage` and `GranuleImageContainer` components we're deleted. The browse images are now implemented as an extension to the `Panels` component. A `EDSCImage` component was also implemented, which creates a standardized way to display a `Spinner` while an image is loading. Only `http` and `https` links with `rel` ending in `/browse#` are supported. 

### What areas of the application does this impact?

The following components we
 - `Map` component
 - `Panels` component

# Testing

### Reproduction steps

- **Environment for testing:** UAT and Prod have better browse imagery
- **Collection to test with:** 
  - Any collection with browse imagery should work
  - C2021957657-LPCLOUD (1 browse image per granule)
  - C1940473819-POCLOUD (4 browse images per granule)
  - C2113473491-NSIDC_ECS (13 browse images per granule)

# Checklist

- [ ] I have added automated tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings

### CODE REVIEW

File: static/src/js/components/GranuleResults/GranuleResultsFocusedMeta.js [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+ * @param {String} props.focusedGranuleMetadata - The metadata for any currently focused granule.
+ * @param {String} props.focusedGranuleId - The id for the focused granule.
+ */
+const GranuleResultsFocusedMeta = ({
+  earthdataEnvironment,
+  focusedGranuleMetadata,
+  focusedGranuleId
+}) => {
+  const { title, links = [], browseFlag } = focusedGranuleMetadata
+  const testProtocol = /^(http|https):\/\//
```
Comment (trevorlang) [inline static/src/js/components/GranuleResults/GranuleResultsFocusedMeta.js:40]: Move this line below

File: static/src/js/components/GranuleResults/GranuleResultsFocusedMeta.js [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+  const [activeModalBrowseImageIndex, setActiveModalBrowseImageIndex] = useState(0)
+  const [browseImageModalIsActive, setBrowseImageModalIsActive] = useState(false)
+  const [showTitleTooltip, setShowTitleTooltip] = useState(false)
+  const [showBrowseImageSelectionPopover, setShowBrowseImageSelectionPopover] = useState(false)
+  const [showModalImageSelectionPopover, setShowModalImageSelectionPopover] = useState(false)
+  const [hideTitleTooltip, setHideTitleTooltip] = useState(false)
+
+  // Filter the links on the granule to find all browse links with an http/https protocol. This filters
+  // any S3 browse links which cause protocol issues.
+  const browseThumbnails = links.filter(({ rel, href }) => rel.indexOf('/browse#') > -1 && testProtocol.test(href))
```
Comment (DanielJDufour) [inline static/src/js/components/GranuleResults/GranuleResultsFocusedMeta.js:51]: `rel.includes('/browse#')`

File: static/src/js/components/Panels/Panels.js [hunk 3]
```diff
@@ -464,6 +464,27 @@ export class Panels extends PureComponent {
...
+      const routeWrapperWidth = routeWrapper.getBoundingClientRect().width
+
+      console.log({
+        width,
+        routeWrapperWidth,
+        thumbnailWidth,
+        eq: width > routeWrapperWidth - thumbnailWidth
+      })
+
+      if (width > routeWrapperWidth - thumbnailWidth - 20) {
```
Comment (trevorlang) [inline static/src/js/components/Panels/Panels.js:482]: Tweak this buffer

File: static/src/js/containers/GranuleResultsFocusedMetaContainer/__tests__/GranuleResultsBodyContainer.test.js
Comment (trevorlang) [inline static/src/js/containers/GranuleResultsFocusedMetaContainer/__tests__/GranuleResultsBodyContainer.test.js:1]: What happened here?

File: static/src/js/containers/MapContainer/MapWrapper.js [hunk 4]
```diff
@@ -10,7 +10,7 @@ import crsProjections from '../../util/map/crs'
 
 import ConnectedSpatialSelectionContainer from '../SpatialSelectionContainer/SpatialSelectionContainer'
 import GranuleGridLayer from '../../components/Map/GranuleGridLayer'
-import GranuleImageContainer from '../GranuleImageContainer/GranuleImageContainer'
```
Comment (trevorlang) [inline static/src/js/containers/MapContainer/MapWrapper.js:13]: Delete this

File: static/src/js/containers/SearchPanelsContainer/SearchPanelsContainer.js
Comment (trevorlang) [inline static/src/js/containers/SearchPanelsContainer/SearchPanelsContainer.js:1]: Revert these changes

File: static/src/js/routes/Search/Search.js [hunk 5]
```diff
@@ -111,6 +112,14 @@ export const Search = ({
     })
   }
 
+  const onUpdateSearchPanelWidth = (width) => {
+    setSearchPanelsWidth(width)
+  }
+
+  useEffect(() => {
```
Comment (trevorlang) [inline static/src/js/routes/Search/Search.js:119]: Delete

File: static/src/js/routes/Search/Search.js
Comment (trevorlang) [inline static/src/js/routes/Search/Search.js:1]: Revert this file

File: static/src/js/components/GranuleResults/GranuleResultsFocusedMeta.js (hunk 2 above)
Comment (trevorlang) [reply static/src/js/components/GranuleResults/GranuleResultsFocusedMeta.js:51]: Smart 😄

File: static/src/js/components/EDSCImage/__tests__/EDSCImage.test.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      const spinner = screen.queryByTestId('edsc-image-spinner')
+
+      expect(image).toBeInTheDocument()
+      expect(container.firstChild.classList.contains('edsc-image--is-loaded')).toEqual(false)
+      expect(spinner).toBeInTheDocument()
+    })
+  })
+
+  describe('when the image has loaded', () => {
+    test('should render a spinner', async () => {
```
Comment (eudoroolivares2016) [inline static/src/js/components/EDSCImage/__tests__/EDSCImage.test.js:31]:
```suggestion
    test('should not render a spinner', async () => {
```

File: static/src/js/components/EDSCImage/__tests__/EDSCImage.test.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      fireEvent.load(image)
+
+      expect(image).toBeInTheDocument()
+      expect(container.firstChild.classList.contains('edsc-image--is-loaded')).toEqual(true)
+      expect(spinner).not.toBeInTheDocument()
+    })
+  })
+
+  describe('when the image has errored', () => {
+    test('should render a spinner', async () => {
```
Comment (eudoroolivares2016) [inline static/src/js/components/EDSCImage/__tests__/EDSCImage.test.js:55]:
```suggestion
    test('should not render a spinner', async () => {
```

File: static/src/js/components/EDSCImage/EDSCImage.scss [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  &__spinner {
+    position: absolute;
+    top: 50%;
+    left: 50%;
+    transform: translate(-50%, -50%);
+    z-index: 1;
+  }
+}
```
Comment (eudoroolivares2016) [inline static/src/js/components/EDSCImage/EDSCImage.scss:29]:
```suggestion
}

```

File: static/src/js/components/EDSCImage/EDSCImage.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * Renders EDSCImage.
+ * @param {Object} props - The props passed into the component.
+ * @param {String} props.alt - The alt text to be used as the alt attribute on the image.
+ * @param {String} props.className - An optional css class attribute.
+ * @param {String} props.height - The height of the image.
+ * @param {String} props.src - The src to be used as the src attribute on the image..
+ * @param {String} props.srcSet - The srcSet to be used as the srcSet attribute on the image.
+ * @param {String} props.width - The width of the imag.
```
Comment (eudoroolivares2016) [inline static/src/js/components/EDSCImage/EDSCImage.js:17]:
```suggestion
 * @param {String} props.width - The width of the image.
```

File: static/src/js/containers/MapContainer/MapWrapper.js [hunk 10]
```diff
@@ -202,7 +201,7 @@ const MapWrapper = (props) => {
         onToggleTooManyPointsModal={onToggleTooManyPointsModal}
         onUpdateShapefile={onUpdateShapefile}
       />
-      <GranuleImageContainer />
+      {/* <GranuleImageContainer /> */}
```
Comment (eudoroolivares2016) [inline static/src/js/containers/MapContainer/MapWrapper.js:204]: Should we remove this since that component is now deleted?

File: static/src/js/components/Panels/Panels.scss [hunk 11]
```diff
@@ -63,4 +63,18 @@
...
+    transition: all 0.2s ease-in-out;
+    opacity: 1;
+  }
+
+  .panels--hide-thumbnail + .panels__focused-meta {
+    visibility: hidden;
+    opacity: 0;
+    pointer-events: none;
+  }
 }
```
Comment (eudoroolivares2016) [inline static/src/js/components/Panels/Panels.scss:80]:
```suggestion
}

```

File: static/src/js/components/GranuleResults/GranuleResultsFocusedMeta.scss [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    }
+  }
+
+  &__modal-nav-primary,
+  &__modal-nav-secondary {
+    display: flex;
+    align-items: center;
+    justify-content: center;
+  }
+}
```
Comment (eudoroolivares2016) [inline static/src/js/components/GranuleResults/GranuleResultsFocusedMeta.scss:185]:
```suggestion
}

```

File: static/src/js/components/GranuleResults/GranuleResultsFocusedMeta.js [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+
+                      return (
+                        href && (
+                          <EDSCImage
+                            key={href}
+                            className={thumbnailClassName}
+                            srcSet={`${imgSrc1x} 1x, ${imgSrc2x} 2x`}
+                            src={imgSrc1x}
+                            alt={description || `Browse image for ${title}`}
+                            width="150px"
```
Comment (eudoroolivares2016) [inline static/src/js/components/GranuleResults/GranuleResultsFocusedMeta.js:286]: For the case where it is `imgSrc2x` does this width& height need to also be set to 300?

File: static/src/js/components/GranuleResults/__tests__/GranuleResultsFocusedMeta.test.js [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        />
+      )
+
+      const focusedMeta = screen.queryByTestId('granule-results-focused-meta')
+
+      expect(focusedMeta).not.toBeInTheDocument()
+    })
+  })
+
+  describe('when links are provided is false', () => {
```
Comment (eudoroolivares2016) [inline static/src/js/components/GranuleResults/__tests__/GranuleResultsFocusedMeta.test.js:51]:
```suggestion
  describe('when links are provided, () => {
```
apologies if I am mistaken and misunderstanding this test

**Review by eudoroolivares2016 — APPROVED** [review]

File: static/src/js/components/GranuleResults/GranuleResultsFocusedMeta.js (hunk 13 above)
Comment (trevorlang) [reply static/src/js/components/GranuleResults/GranuleResultsFocusedMeta.js:286]: Nope, it will still be displayed at 150x150

File: static/src/js/components/EDSCImage/EDSCImage.js (hunk 9 above)
Comment (trevorlang) [reply static/src/js/components/EDSCImage/EDSCImage.js]: Nice catch. Surprised this slipped by the spell checker :shaking-fist:

**Review by jaybarra — APPROVED** [review]

**Review by DanielJDufour — APPROVED** [review]
